Mage.Sets/src/mage/cards/t/TheTaleOfTamiyo.java

```diff
@@ -0,0 +1,207 @@
+
+                // Check if there's at least one type in common
+                boolean typesMatch = false;
+
+                // Iterate over the second card's types to see if there is any overlap with the first card's types
+                for (CardType type : cards.get(1).getCardType()) {
+                    if (firstCardTypes.contains(type)) {
+                        typesMatch = true;
+                        break;
+                    }
+                }
+
+                // If there is a match, draw a card and continue; otherwise, break
+                if (typesMatch) {
+                    controller.drawCards(1, source, game);
+                } else {
+                    break;
+                }
+
+            } while (controller.getLibrary().size() >= 2); // Continue if there are at least 2 cards left in the library
+
+            return true;
+        }
+        return false;
+    }
+
+
+
+}
+class TheTaleOfTamiyoEffect2 extends OneShotEffect {
+
+    private final FilterCard filter;
+
+    TheTaleOfTamiyoEffect2() {
+        super(Outcome.PlayForFree);
+        this.staticText = "Exile any number of target instant, sorcery, and/or Tamiyo planeswalker cards " +
+                "from your graveyard. Copy them. You may cast any number of the copies.";
+
+        // Initialize filter
+        this.filter = new FilterCard("instant, sorcery, or Tamiyo planeswalker card");
+        this.filter.add(Predicates.or(
+                CardType.INSTANT.getPredicate(),
+                CardType.SORCERY.getPredicate(),
+                SubType.TAMIYO.getPredicate()
+        ));
+    }
+
+    private TheTaleOfTamiyoEffect2(final TheTaleOfTamiyoEffect2 effect) {
+        super(effect);
+        this.filter = effect.filter;  // Copy filter
+    }
+
+    @Override
+    public TheTaleOfTamiyoEffect2 copy() {
+        return new TheTaleOfTamiyoEffect2(this);
+    }
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        if (controller != null) {
+            // Use filter to choose the cards to exile
+            Set<Card> cardsToExile = controller.getGraveyard().getCards(
+                    filter, source.getControllerId(), source, game);
+
+            if (!cardsToExile.isEmpty()) {
+                // Move cards from graveyard to exile
+                if (controller.moveCards(cardsToExile, Zone.EXILED, source, game)) {
+                    Cards copiedCards = new CardsImpl();
+                    // Copy cards
+                    for (Card card : cardsToExile) {
+                        copiedCards.add(game.copyCard(card, source, source.getControllerId()));
+                    }
+
+                    // Cast copies
+                    boolean continueCasting = true;
+                    while (controller.canRespond() && continueCasting && !copiedCards.isEmpty()) {
```

Comment:
Do not use a while loop for this. Instead, use Integer.MAX_VALUE for the max number of targets in the target card, so all the cards to cast can be selected at once. Then for the casting process iterate through those targets with a for loop.